keps/sig-apps/4650-stateful-set-update-claim-template/README.md

```diff
@@ -0,0 +1,1087 @@
+* Validate the updated `volumeClaimTemplates` as how PVC patch does.
+* Update ephemeral volumes.
+* Patch PVCs that are different from the template, e.g. StatefulSet adopts the pre-existing PVCs.
+
+
+## Proposal
+
+<!--
+This is where we get down to the specifics of what the proposal actually is.
+This should have enough detail that reviewers can understand exactly what
+you're proposing, but should not include things like API designs or
+implementation. What is the desired outcome and how do we measure success?.
+The "Design Details" section below is for the real
+nitty-gritty.
+-->
+1. Change API server to allow specific updates to `volumeClaimTemplates` of a StatefulSet:
+   * `labels`
+   * `annotations`
+   * `resources.requests.storage`
+   * `volumeAttributesClassName`
+
+2. Modify StatefulSet controller to add PVC reconciliation logic.
+
+3. Collect the status of managed PVCs, and show them in the StatefulSet status.
+
+### Kubernetes API Changes
+
+Changes to StatefulSet `spec`:
+
+Introduce a new field in StatefulSet `spec`: `volumeClaimUpdatePolicy` to
+specify how to coordinate the update of PVCs and Pods. Possible values are:
+- `OnDelete`: the default value, only update the PVC when the the old PVC is deleted.
+- `InPlace`: patch the PVC in-place if possible. Also includes the `OnDelete` behavior.
+
+Changes to StatefultSet `status`:
+
+Additionally collect the status of managed PVCs, and show them in the StatefulSet status.
+
+For each PVC in the template:
+- compatible: the number of PVCs that are compatible with the template.
+  These replicas will not be blocked on Pod recreation.
+- updating: the number of PVCs that are being updated in-place (e.g. expansion in progress).
+- overSized: the number of PVCs that are larger than the template.
+- totalCapacity: the sum of `status.capacity` of all the PVCs.
+
+Some fields in the `status` are also updated to reflect the staus of the PVCs:
+- readyReplicas: in addition to pods, also consider the PVCs status. A PVC is not ready if:
+  - `volumeClaimUpdatePolicy` is `InPlace` and the PVC is updating;
+- availableReplicas: total number of replicas of which both Pod and PVCs are ready for at least `minReadySeconds`
+- currentRevision, updateRevision, currentReplicas, updatedReplicas
+  are updated to reflect the status of PVCs.
+
+With these changes, user can still use `kubectl rollout status` to monitor the update process,
+both for automated patching and for the PVCs that need manual intervention.
+
+### Updated Reconciliation Logic
+
+How to update PVCs:
+1. If `volumeClaimUpdatePolicy` is `InPlace`,
+   and if `volumeClaimTemplates` and actual PVC only differ in mutable fields
+   (`spec.resources.requests.storage`, `spec.volumeAttributesClassName`, `metadata.labels`, and `metadata.annotations` currently),
+   patch the PVC to the extent possible.
+   - `spec.resources.requests.storage` is patched to max(template spec, PVC status)
```

Comment:
Hi,
I appreciate your work.
It says `max(template spec, PVC status)` here, but isn't it `max(template spec, PVC spec)`? If we use `PVC status`, it might interrupt the user's PVC expansion as follows.

1. Initial state: PVC template: 3G, PVC spec: 3G, PVC status 3G
2. User updates PVC spec: PVC template: 3G, PVC spec: 5G, PVC status: 3G
3. User updates PVC template before csi driver expand volume: PVC template: 4G, PVC spec: 5G, PVC status: 3G
4. The PVC expansion in step 2 won't be reflected: PVC template: 4G, PVC spec: 4G, PVC status: 3G

What do you think?

Reply:
I think this is the acceptable behavior. I see your link to pvc-autoresizer. For a controller to work with this KEP, it should keep trying to reconcile the PVC spec. So, it should update PVC spec again to 5G after step 4. And STS controller will not interference again because it only updates PVC if the `controller-revision-hash` label on the pod does not match.

For a human operator, he should monitor the PVC status to ensure his operation is successful anyway.

This is hard to avoid, I think. I would like to take advantage of KEP-1790 to help recover from expansion failure. So, I allow shrink the PVC spec (not tested yet, not sure how they work together).

Reply:
Thank you for the explanation. I understood it, and I think it will work well now. When the alpha version is released, I will check to see if there are any cases I have overlooked.